File: spring-rest-configuration-autodiscover.java/src/main/java/demo/service/core/PropertyEnumService.java
package demo.service.core;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public abstract class PropertyEnumService<T extends Enum<T>> extends AbstractPropertyService<T> {

    @Override
    public Type getType() {
        return Type.ENUM;
    }

    @Override
    public T fromString(String input) {
        return input == null ? getDefaultValue() : Enum.valueOf(getEnumClass(), input);
    }

    @Override
    public String toString(T input) {
        return (input == null) ? getDefaultValue().toString() : String.valueOf(input);
    }

    @Override
    public List<String> getPossibleValues() {
        return Stream.of(getEnumClass().getEnumConstants()).map(Enum::name).collect(Collectors.toList());
    }

    protected abstract Class<T> getEnumClass();
}

File: spring-rest-configuration-autodiscover.groovy/src/main/kotlin/demo/service/core/PropertyEnumService.kt
package demo.service.core

abstract class PropertyEnumService<T : Enum<T>> : AbstractPropertyService<T>() {

    override fun getType() = Type.ENUM

    public override fun fromString(input: String?): T =
            if (input == null) getDefaultValue() else java.lang.Enum.valueOf<T>(getEnumClass(), input)

    public override fun toString(input: T?): String =
            input?.toString() ?: getDefaultValue().toString()

    override fun getPossibleValues(): List<String> = getEnumClass().enumConstants.map { it -> it.name }

    protected abstract fun getEnumClass(): Class<T>
}
